src/dbus_api/blockdev.rs

```diff
@@ -0,0 +1,276 @@
+                                parent: dbus::Path<'static>,
+                                uuid: Uuid)
+                                -> dbus::Path<'a> {
+    let f = Factory::new_fn();
+
+    let set_userid_method = f.method("SetUserInfo", (), set_user_info)
+        .in_arg(("id", "s"))
+        .out_arg(("changed", "b"))
+        .out_arg(("return_code", "q"))
+        .out_arg(("return_string", "s"));
+
+    let devnode_property = f.property::<&str, _>("Devnode", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::Const)
+        .on_get(get_blockdev_devnode);
+
+    let hardware_info_property = f.property::<&str, _>("HardwareInfo", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::Const)
+        .on_get(get_blockdev_hardware_info);
+
+    let user_info_property = f.property::<&str, _>("UserInfo", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::False)
+        .on_get(get_blockdev_user_info);
+
+    let initialization_time_property = f.property::<&str, _>("InitializationTime", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::Const)
+        .on_get(get_blockdev_initialization_time);
+
+    let total_physical_size_property = f.property::<&str, _>("TotalPhysicalSize", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::False)
+        .on_get(get_blockdev_physical_size);
+
+    let state_property = f.property::<u16, _>("State", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::False)
+        .on_get(get_blockdev_state);
+
+    let pool_property = f.property::<&dbus::Path, _>("Pool", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::Const)
+        .on_get(get_parent);
+
+    let uuid_property = f.property::<&str, _>("Uuid", ())
+        .access(Access::Read)
+        .emits_changed(EmitsChangedSignal::Const)
+        .on_get(get_uuid);
+
+    let object_name = format!("{}/{}",
+                              STRATIS_BASE_PATH,
+                              dbus_context.get_next_id().to_string());
+
+    let interface_name = format!("{}.{}", STRATIS_BASE_SERVICE, "blockdev");
+
+    let object_path = f.object_path(object_name, Some(OPContext::new(parent, uuid)))
+        .introspectable()
+        .add(f.interface(interface_name, ())
+                 .add_m(set_userid_method)
+                 .add_p(devnode_property)
+                 .add_p(hardware_info_property)
+                 .add_p(initialization_time_property)
+                 .add_p(total_physical_size_property)
+                 .add_p(pool_property)
+                 .add_p(state_property)
+                 .add_p(user_info_property)
+                 .add_p(uuid_property));
+
+    let path = object_path.get_name().to_owned();
+    dbus_context.actions.borrow_mut().push_add(object_path);
+    path
+}
+
+fn set_user_info(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
```

Comment:
How could this method ever allow setting user info to None. Is that desired?

Reply:
Well I think so. The user may attach user info to a device, and then later want to remove what they added. Of course they could set it to an empty string but that's not ideal. Is there a way we can make it possible to unset the info using this method? Or do we unfortunately need an `clear_user_info` method as well?

Reply:
Hmm, in `get_blockdev_user_info` we convert None to Some(""). This suggests another alternative would be doing the opposite in `set`. I guess my main goal is to avoid having `user_info = ""` in the on-disk json.